json/pickler/src/main/scala/sttp/tapir/json/pickler/Pickler.scala

```diff
@@ -0,0 +1,368 @@
+    val tpe = TypeRepr.of[T]
+
+    val isValueCaseClass =
+      tpe.typeSymbol.isClassDef && tpe.classSymbol.get.flags.is(Flags.Case) && tpe.baseClasses.contains(Symbol.classSymbol("scala.AnyVal"))
+
+    if (!isValueCaseClass) {
+      '{ nonMirrorPickler[T] }
+    } else {
+
+      val field = tpe.typeSymbol.declaredFields.head
+      val fieldTpe = tpe.memberType(field)
+      fieldTpe.asType match
+        case '[f] =>
+          val basePickler = Expr.summon[Pickler[f]].getOrElse {
+            report.errorAndAbort(
+              s"Cannot summon Pickler for value class ${tpe.show}. Missing Pickler[${fieldTpe.show}] in implicit scope."
+            )
+          }
+          '{
+            val newSchema: Schema[T] = ${ basePickler }.schema.as[T]
+            new Pickler[T](
+              new TapirPickle[T] {
+                override lazy val writer = summonInline[Writer[f]].comap[T](
+                  // writing object of type T means writing T.field
+                  ccObj => ${ Select.unique(('ccObj).asTerm, field.name).asExprOf[f] }
+                )
+                // a reader of type f (field) will read it and wrap into value object using the consutructor of T
+                override lazy val reader = summonInline[Reader[f]]
+                  .map[T](fieldObj => ${ Apply(Select.unique(New(Inferred(tpe)), "<init>"), List(('fieldObj).asTerm)).asExprOf[T] })
+              },
+              newSchema
+            )
+          }
+    }
+
+  private inline def fromExistingSchemaAndRw[T](schema: Schema[T])(using ClassTag[T], Configuration, Mirror.Of[T]): Pickler[T] =
+    Pickler(
+      new TapirPickle[T] {
+        val rw: ReadWriter[T] = summonFrom {
+          case foundTapirRW: ReadWriter[T] =>
+            foundTapirRW
+          case foundUpickleDefaultRW: _root_.upickle.default.ReadWriter[T] => // there is BOTH schema and ReadWriter in scope
+            foundUpickleDefaultRW.asInstanceOf[ReadWriter[T]]
+          case _ =>
+            errorForType[T](
+              "Found implicit Schema[%s] but couldn't find a uPickle ReadWriter for this type. Either provide a ReadWriter, or remove the Schema from scope and let Pickler derive its own."
+            )
+            null
+        }
+        override lazy val reader = rw
+        override lazy val writer = rw
+      },
+      schema
+    )
+
+  private[pickler] inline def buildNewPickler[T: ClassTag](
+  )(using m: Mirror.Of[T], c: Configuration, subtypeDiscriminator: SubtypeDiscriminator[T]): Pickler[T] =
+    // The lazy modifier is necessary for preventing infinite recursion in the derived instance for recursive types such as Lst
+    lazy val childPicklers: Tuple.Map[m.MirroredElemTypes, Pickler] = summonChildPicklerInstances[T, m.MirroredElemTypes]
+    inline m match {
+      case p: Mirror.ProductOf[T] => picklerProduct(p, childPicklers)
+      case _: Mirror.SumOf[T] =>
+        val schema: Schema[T] =
+          inline if (isScalaEnum[T])
```

Comment:
is this enough? that is, won't this also catch cases of scala 3 enums which have parameters, in which case they desugar to a sealed trait hierarchy?

Reply:
This indeed catches enums with fields. This means, that:

```scala
    // given
    enum RichColorEnum(val code: Int):
      case Cyan extends RichColorEnum(3)
      case Magenta extends RichColorEnum(18)

    case class RichColorResponse(color: RichColorEnum)

    // when
    val picklerResponse = Pickler.derived[RichColorResponse]
    val codec = picklerResponse.toCodec
    val inputObj = RichColorResponse(RichColorEnum.Cyan)
    val encoded = codec.encode(inputObj)

    // then
    encoded shouldBe """{"color":"Cyan"}"""
```

Would you rather expect the `code` field to be encoded as well? If someone used `enum` to represent this, then I would expect it to be treated as enumeration with simple values in the schema and JSON model.

Reply:
Here - no, as the enum cases themselves are not parametrised. But you can have 

```scala
enum Entity:
  case Person(first: String, age: Int)
  case Business(address: String)
```

which desugars to a sealed trait + case class hierarchy. In essence scala 3 enums have two flavors: when all cases are parameterless, then you have an enumeration of values, and you can represent this using just the label. (Also, the compiler won't generate a separate type for such cases). Or, if you have parameters, then this is syntax sugar for sealed trait + case classes. Both need to be treated differently: the first should be handled using `derivedEnumeration`, the second as a coproduct (sum type), using inheritance strategies etc.

I think we have this kind of check somewhere in the code, maybe in `Schema.derivedEnumeration`? For sure it mentions parameterlessness in the docs, not sure if there's an actual check.

Reply:
Btw would be great to have a test case for each of those (at least 3 - very simple enums, parameterless enums with base type parameters, parametrised enums). The existing schema tests might be messy, so maybe it would make sense to have some separate tests as well

Reply:
Interesting, thanks for explaining. I don't like this idea of allowing writing sealed hierarchies as enums, seems like unnecessary facilitation of already simple thing, and makes it unlcear whether it's an enum or "not really an enum". There's indeed a nice check you've implemented in `Validator.derivedEnumeration[T]` which fails compilation for such cases.
I can add this check to our `isScalaEnum` utility and return false, which would make the logic go through a standard product path and produce
```
{ "$type": "Business", "address": "221B Baker Street" }
```

Reply:
This should do the trick then https://github.com/softwaremill/tapir/pull/3134/commits/f5879e75c7ea9dfb5623a64c32d4d6c114e1997f